### ISSUE TITLE
NamespaceSpec and ObjectMeta both contain a "finalizers" field

### ISSUE THREAD

**Issue by khogeland** [kubernetes/kubernetes#65994]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

`Namespace` has two finalizers fields: `Namespace.ObjectMeta.finalizers` and `Namespace.Spec.finalizers`. `NamespacedResourcesDeleter` uses `NamespaceSpec.finalizers` and `ObjectMeta.finalizers` appears to be unused/ineffective (please correct me if I'm wrong).

**What you expected to happen**:

`NamespacedResourcesDeleter` should use `ObjectMeta.finalizers`, or if there is a technical reason for not doing so, it should be documented how and why the fields are different. Also, if the `ObjectMeta.finalizers` field is completely unused for namespaces (can't quite tell if it is), maybe the API should fail namespace creation requests that set it.

TL;DR: I mistakenly put a finalizer in the namespace metadata, instead of the namespace spec. That should either be supported, obviously wrong via documentation/naming, or not be a valid request.

**Comment by yue9944882** [comment]
Hi, @khogeland `ObjectMeta.finalizers` is used for generic use and these finalizers will be processed by generic store logic and gc mechanism which is inside the apiserver. As for `NamespaceSpec.finalizers` is dedicated for namespace controller which is in the another component. As you can see they are processed by separated controllers so it's questionable to merge these finalizers into the one common field. 😃It's not really a BUG AFAICT.

**Comment by caesarxuchao** [comment]
>  I mistakenly put a finalizer in the namespace metadata, instead of the namespace spec. That should either be supported, obviously wrong via documentation/naming, or not be a valid request.

We can't tighten the validation, since that would be backwards incompatible.

We can update the namespace registry to respect metadata.finalizer as well.

**Comment by nikhita** [comment]
> We can't tighten the validation, since that would be backwards incompatible.

maybe a use case for ratcheting validation: https://github.com/kubernetes/kubernetes/pull/64907?

/cc @liggitt

**Comment by liggitt** [comment]
>> We can't tighten the validation, since that would be backwards incompatible.
>
> maybe a use case for ratcheting validation: #64907?

No, both API fields have meaning and can be used to good effect:
* `metadata.finalizers` can be populated just like any other object, and dealt with consistently. We would not break this.
* `spec.finalizers` has special access patterns for namespaces (the `/finalize` subresource), which we also would not break.

I think the action items here are:
* fix the namespace registry custom Delete implementation to populate `metadata.finalizers` with garbage collection related finalizers depending on the incoming delete options
* improve the documentation of the `spec.finalizers` field to clarify the difference between it and `metadata.finalizers` and when each should be used.

**Comment by lavalamp** [comment]
Just came across this-- I actually do consider the current state a big bug, but it's not fixable due to api version change rules.
